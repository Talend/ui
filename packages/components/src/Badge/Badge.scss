$tc-badge-label-font-size: 13px !default;
$tc-badge-height: 25px !default;
$tc-badge-delete-icon-size: $svg-xs-size !default;
// it should sum up in 25px height:
$tc-badge-vertical-padding: ($tc-badge-height - $tc-badge-label-font-size * $line-height-base) / 2 !default;
$tc-badge-margin: $padding-smaller !default;

.tc-badge {
	display: inline-flex;
	max-width: 100%;

	.tc-badge-button {
		display: inline-flex;
		height: $tc-badge-height;
		margin: $tc-badge-margin;
		padding: $tc-badge-vertical-padding $padding-small;
		padding-right: calc(#{$padding-larger} + #{$tc-badge-delete-icon-size});
		background-color: $gallery;
		letter-spacing: 0.4px;
		border: none;
		max-width: 100%;

		.tc-badge-category {
			background-color: $concrete;
			font-size: $tc-badge-label-font-size;
			text-transform: uppercase;

			margin-top: -$tc-badge-vertical-padding;
			margin-right: $padding-small;
			margin-left: -$padding-small;
			margin-bottom: -$tc-badge-vertical-padding;
			padding: $tc-badge-vertical-padding $padding-small;
		}

		.tc-badge-label {
			font-size: $tc-badge-label-font-size;
			font-weight: 700;

			overflow: hidden;
			text-overflow: ellipsis;
			white-space: nowrap;
		}

		&:hover {
			background-color: $alto;

			.tc-badge-category {
				background-color: $gallery;
			}
		}
	}

	svg,
	.tc-badge-label,
	.tc-badge-category {
		color: $dark-silver;
	}

	&-readonly.tc-badge {
		.tc-badge-button {
			padding-right: $padding-small;
		}
	}

	&-selected:not(&-disabled) {
		svg,
		.tc-badge-label,
		.tc-badge-category {
			color: $white;
		}

		.tc-badge-button {
			background-color: $scooter;

			.tc-badge-category {
				background-color: rgba($white, 0.2);
			}

			&:hover {
				background-color: darken($scooter, 10);
			}
		}
	}

	&-disabled {
		.tc-badge-button {
			background-color: $gallery;

			svg,
			.tc-badge-label,
			.tc-badge-category {
				color: $silver;
			}

			.tc-badge-category {
				background-color: $concrete;
			}

			& + span {
				display: flex;
			}
		}
	}

	.tc-badge-delete-icon {
		position: relative;
		margin: auto;
		height: $tc-badge-height - 2 * $tc-badge-vertical-padding;
		padding: 0;
		background: transparent;
		box-shadow: none;
		left: -$padding-large;

		svg {
			display: block;
			margin: 0;
			height: $tc-badge-delete-icon-size;
			width: $tc-badge-delete-icon-size;
		}
	}
}
