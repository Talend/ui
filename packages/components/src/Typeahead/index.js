import Typeahead from './Typeahead.component';

export default Typeahead;
