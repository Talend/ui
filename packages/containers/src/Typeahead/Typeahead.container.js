import React from 'react';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import { cmfConnect, componentState } from '@talend/react-cmf';
import { Typeahead as Component } from '@talend/react-components';
import omit from 'lodash/omit';
import keycode from 'keycode';

export const DISPLAY_NAME = 'Container(Typeahead)';
export const DEFAULT_STATE = new Immutable.Map({
	docked: true,
	searching: false,
	focusedSectionIndex: null,
	focusedItemIndex: null,
	items: null,
});

/**
 * The Typeahead React container
 */
export default class Typeahead extends React.Component {
	static displayName = DISPLAY_NAME;
	static propTypes = {
		...componentState.propTypes,
		onSelect: PropTypes.func,
		onChange: PropTypes.func,
		onBlur: PropTypes.func,
	};

	constructor(props) {
		super(props);
		this.onToggle = this.onToggle.bind(this);
		this.onKeyDown = this.onKeyDown.bind(this);
		this.onBlur = this.onBlur.bind(this);
		this.onSelect = this.onSelect.bind(this);
	}

	onToggle() {
		this.props.setState(() => ({
			docked: !this.props.state.get('docked', true),
			focusedSectionIndex: null,
			focusedItemIndex: null,
			items: null,
		}));
	}

	onBlur(event) {
		const { onBlur } = this.props;

		this.onToggle();

		if (onBlur) {
			onBlur(event);
		}
	}

	onSelect(event, value) {
		const { onSelect } = this.props;

		if (onSelect) {
			onSelect(event, value);
		}
	}

	onKeyDown(event, data) {
		const { onKeyDown } = this.props;
		const {
			highlightedItemIndex,
			newHighlightedItemIndex,
			highlightedSectionIndex,
			newHighlightedSectionIndex,
		} = data;

		if (onKeyDown) {
			onKeyDown(event, data);
		}
		switch (event.which) {
			case keycode.codes.down:
			case keycode.codes.up:
				event.preventDefault();
				this.props.setState(() => ({
					focusedSectionIndex: newHighlightedSectionIndex,
					focusedItemIndex: newHighlightedItemIndex,
				}));
				break;
			case keycode.codes.enter:
				event.preventDefault();
				if (highlightedItemIndex !== null && highlightedItemIndex !== null) {
					this.onSelect(event, {
						sectionIndex: highlightedSectionIndex,
						itemIndex: highlightedItemIndex,
					});
				}
				break;
			case keycode.codes.esc:
				event.preventDefault();
				this.onBlur(event);
				break;
			default:
		}
	}

	render() {
		const { items } = this.props;
		const props = {
			...omit(this.props, cmfConnect.INJECTED_PROPS),
			...this.props.state.toJS(),
			onToggle: this.onToggle,
			onBlur: this.onBlur,
			onSelect: this.onSelect,
			onKeyDown: this.onKeyDown,
			items: items && items.toJS ? items.toJS() : items,
		};

		return <Component {...props} />;
	}
}
