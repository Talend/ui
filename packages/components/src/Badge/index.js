import Badge from './Badge.component';

export default Badge;
