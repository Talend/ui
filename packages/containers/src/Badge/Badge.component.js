import { cmfConnect } from '@talend/react-cmf';
import Badge from '@talend/react-components/lib/Badge';

export default cmfConnect({
	omitCMFProps: true,
	withComponentRegistry: true,
	withDispatch: true,
	withDispatchActionCreator: true,
	withComponentId: true,
})(Badge);
