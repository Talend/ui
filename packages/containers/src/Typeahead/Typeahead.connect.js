import { cmfConnect } from '@talend/react-cmf';
import Typeahead, { DEFAULT_STATE } from './Typeahead.container';

export default cmfConnect({
	defaultState: DEFAULT_STATE,
	omitCMFProps: true,
	withComponentRegistry: true,
	withDispatch: true,
	withDispatchActionCreator: true,
	withComponentId: true,
})(Typeahead);
