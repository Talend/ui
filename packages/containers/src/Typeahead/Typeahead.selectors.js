import { DEFAULT_STATE, DISPLAY_NAME } from './Typeahead.container';

/**
 * Selector on the state from the typeahead id given.
 * @param {object} state
 * @param {string} id
 */
export function getComponentState(state, id) {
	return state.cmf.components.getIn([DISPLAY_NAME, id], DEFAULT_STATE);
}
