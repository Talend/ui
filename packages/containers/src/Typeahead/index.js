import Typeahead from './Typeahead.connect';

export default Typeahead;
