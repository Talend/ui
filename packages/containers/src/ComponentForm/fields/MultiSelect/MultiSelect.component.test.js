import React from 'react';
import ReactDOM from 'react-dom';
import { mount, shallow } from 'enzyme';
import Badge from '@talend/react-components/lib/Badge';
import Typeahead from '@talend/react-components/lib/Typeahead';
import keycode from 'keycode';
import MultiSelectTag from './MultiSelectTag.component';

describe('MultiSelectTag field', () => {
	const props = {
		id: 'my-select-tag',
		isValid: true,
		errorMessage: 'This is wrong',
		onChange: jest.fn(),
		onFinish: jest.fn(),
		schema: {
			autoFocus: true,
			description: 'This is the MultiSelectTag field',
			disabled: false,
			placeholder: 'Type here',
			readOnly: false,
			required: true,
			restricted: false,
			title: 'Tags',
			titleMap: [{ name: 'toto', value: 'titi' }, { name: 'tata', value: 'tutu' }],
		},
		value: ['aze', 'tutu'],
	};

	it('should render MultiSelectTag', () => {
		// when
		const wrapper = shallow(<MultiSelectTag {...props} />);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should update suggestion on input change', () => {
		// given
		const wrapper = mount(<MultiSelectTag {...props} />);

		// when
		wrapper
			.find('input')
			.at(0)
			.simulate('change', { target: { value: 'titi' } });

		// then
		expect(wrapper.find(Typeahead).props().items).toEqual([{ title: 'titi (new)', value: 'titi' }]);
	});

	it('should update suggestion on props.value change', () => {
		// given
		const node = document.createElement('div');
		// eslint-disable-next-line react/no-render-return-value
		const instance = ReactDOM.render(<MultiSelectTag {...props} />, node);
		instance.updateSuggestions();
		expect(instance.state.suggestions).toEqual([{ title: 'toto', value: 'titi' }]);

		// when : trigger a props update
		ReactDOM.render(<MultiSelectTag {...props} value={['aze']} />, node);

		// then
		expect(instance.state.suggestions).toEqual([
			{ title: 'toto', value: 'titi' },
			{ title: 'tata', value: 'tutu' },
		]);
	});

	it('should suggest new item creation when widget is not restricted', () => {
		// given
		const wrapper = mount(<MultiSelectTag {...props} />);

		// when
		wrapper
			.find('input')
			.at(0)
			.simulate('change', { target: { value: 'az' } });

		// then
		expect(wrapper.find(Typeahead).props().items).toEqual([{ title: 'az (new)', value: 'az' }]);
	});

	it('should NOT suggest new item creation when widget is restricted', () => {
		// given
		const restrictedSchema = { ...props.schema, restricted: true };
		const wrapper = mount(<MultiSelectTag {...props} schema={restrictedSchema} />);

		// when
		wrapper
			.find('input')
			.at(0)
			.simulate('change', { target: { value: 'az' } });

		// then
		expect(wrapper.find(Typeahead).props().items).toEqual([]);
	});

	it('should add tag', () => {
		// given
		const onChange = jest.fn();
		const onFinish = jest.fn();
		const wrapper = mount(<MultiSelectTag {...props} onChange={onChange} onFinish={onFinish} />);
		const input = wrapper.find('input').at(0);

		// when
		input.simulate('change', { target: { value: 'titi' } });
		input.simulate('keydown', { which: keycode.codes.enter });

		// then
		const payload = { schema: props.schema, value: props.value.concat('titi') };
		expect(onChange).toBeCalledWith(expect.anything(), payload);
		expect(onFinish).toBeCalledWith(expect.anything(), payload);
	});

	it('should remove tag', () => {
		// given
		const onChange = jest.fn();
		const onFinish = jest.fn();
		const wrapper = mount(<MultiSelectTag {...props} onChange={onChange} onFinish={onFinish} />);

		// when
		wrapper
			.find('.tc-badge-delete-icon')
			.at(0)
			.simulate('click');

		// then
		const payload = { schema: props.schema, value: props.value.slice(1) };
		expect(onChange).toBeCalledWith(expect.anything(), payload);
		expect(onFinish).toBeCalledWith(expect.anything(), payload);
	});

	it('should call onTrigger on focus', done => {
		// given
		const data = { titleMap: [{ name: 'Foo', value: 'foo' }] };
		const triggerProps = {
			...props,
			onTrigger: jest.fn(
				event =>
					new Promise(resolve => {
						// hack: to be sure we catch the setState after the promise
						setTimeout(() => {
							expect(event.target.state.isLoading).toBe(false);
							done();
						}, 0);
						return resolve(data);
					}),
			),
			schema: {
				...props.schema,
				triggers: [
					{
						onEvent: 'focus',
					},
				],
			},
		};
		const wrapper = shallow(<MultiSelectTag {...triggerProps} />);
		const event = { type: 'focus', target: wrapper.instance() };

		// when
		wrapper
			.find('FieldTemplate')
			.find(Typeahead)
			.prop('onFocus')(event);

		// then
		expect(triggerProps.onTrigger).toBeCalledWith(event, {
			trigger: triggerProps.schema.triggers[0],
			schema: triggerProps.schema,
			errors: triggerProps.errors,
			properties: triggerProps.properties,
		});
		expect(wrapper.state('isLoading')).toBe(true);
	});

	it('should call onTrigger on focus', () => {
		// given
		const nameResolverProps = {
			...props,
			resolveName: value => value.map(next => `${next}_name`),
		};
		const wrapper = shallow(<MultiSelectTag {...nameResolverProps} />);

		// when
		const firstLabel = wrapper
			.find('FieldTemplate')
			.find(Badge)
			.first()
			.prop('label');

		// then
		expect(firstLabel).toBe('aze_name');
	});
});
