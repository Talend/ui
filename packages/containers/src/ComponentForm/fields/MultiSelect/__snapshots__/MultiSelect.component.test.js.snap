// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect field should render MultiSelect 1`] = `
<FieldTemplate
  description="This is the MultiSelect field"
  descriptionId="my-select-tag-description"
  errorId="my-select-tag-error"
  errorMessage="This is wrong"
  id="my-select-tag"
  isValid={false}
  label="Tags"
  required={true}
>
  <Translate(MultiSelect)
    autoFocus={true}
    disabled={false}
    id="my-select-tag"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    options={
      Array [
        Object {
          "name": "toto",
          "value": "titi",
        },
        Object {
          "name": "tata",
          "value": "tutu",
        },
      ]
    }
    placeholder="Type here"
    readOnly={false}
    required={true}
    selected={
      Array [
        "aze",
        "tutu",
      ]
    }
    withCreateNew={true}
  />
</FieldTemplate>
`;
